File: app/src/main/java/com/websarva/wings/android/zuboradiary/ui/list/WordSearchResultDayListAdapter.java
package com.websarva.wings.android.zuboradiary.ui.list;

import android.content.Context;
import android.text.SpannableString;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.navigation.NavDirections;
import androidx.recyclerview.widget.DiffUtil;
import androidx.recyclerview.widget.ListAdapter;
import androidx.recyclerview.widget.RecyclerView;

import com.websarva.wings.android.zuboradiary.R;
import com.websarva.wings.android.zuboradiary.data.DateConverter;
import com.websarva.wings.android.zuboradiary.databinding.RowDiaryDayListBinding;
import com.websarva.wings.android.zuboradiary.databinding.RowWordSearchResultBinding;
import com.websarva.wings.android.zuboradiary.ui.list.wordsearch.WordSearchFragment;
import com.websarva.wings.android.zuboradiary.ui.list.wordsearch.WordSearchFragmentDirections;
import com.websarva.wings.android.zuboradiary.ui.list.wordsearch.WordSearchResultDayListItem;

import java.time.LocalDate;
import java.util.function.Consumer;

public class WordSearchResultDayListAdapter
        extends ListAdapter<WordSearchResultDayListItem, WordSearchResultDayListAdapter.WordSearchResultDayViewHolder> {

    private final Context context;
    private final Consumer<LocalDate> processOnClick;

    public WordSearchResultDayListAdapter(Context context, Consumer<LocalDate> processOnClick){
        super(new WordSearchResultDayListDiffUtilItemCallback());
        this.context = context;
        this.processOnClick = processOnClick;
    }

    //日記リスト(日)のホルダーと日記リスト(日)のアイテムレイアウトを紐づける。
    @NonNull
    @Override
    public WordSearchResultDayViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        RowWordSearchResultBinding binding =
                RowWordSearchResultBinding
                        .inflate(LayoutInflater.from(parent.getContext()), parent, false);
        return new WordSearchResultDayViewHolder(binding);
    }

    //日記リスト(日)の各行アイテム(ホルダー)情報を設定。
    @Override
    public void onBindViewHolder(WordSearchResultDayViewHolder holder, int position) {
        WordSearchResultDayListItem item = getItem(position);
        int year = item.getYear();
        int month = item.getMonth();
        int dayOfMonth = item.getDayOfMonth();
        String dayOfWeek = item.getDayOfWeek();
        SpannableString title = item.getTitle();
        int itemNumber = item.getItemNumber();
        SpannableString itemTitle = item.getItemTitle();
        SpannableString itemComment = item.getItemComment();
        holder.date = DateConverter.toStringLocalDate(year, month, dayOfMonth); // ホルダー毎に日記の日付情報一式付与
        holder.textDayOfMonth.setText(dayOfWeek);
        holder.textDayOfMonth.setText(String.valueOf(dayOfMonth));
        holder.textWordSearchResultTitle.setText(title);
        String stringItemNumber = "項目" + String.valueOf(itemNumber);
        holder.textWordSearchResultItemNumber.setText(stringItemNumber);
        holder.textWordSearchResultItemTitle.setText(itemTitle);
        holder.textWordSearchResultItemComment.setText(itemComment);
        holder.itemView.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                processOnClick.accept();
            }
        });
    }

    public static class WordSearchResultDayViewHolder extends RecyclerView.ViewHolder {

        private final RowWordSearchResultBinding binding;
        private LocalDate date;

        public WordSearchResultDayViewHolder(RowWordSearchResultBinding binding) {
            super(binding.getRoot());
            this.binding = binding;
        }
    }

    public static class WordSearchResultDayListDiffUtilItemCallback
            extends DiffUtil.ItemCallback<WordSearchResultDayListItem> {
        @Override
        public boolean areItemsTheSame(@NonNull WordSearchResultDayListItem oldItem, @NonNull WordSearchResultDayListItem newItem) {
            return oldItem.getId().equals(newItem.getId());
        }

        @Override
        public boolean areContentsTheSame(@NonNull WordSearchResultDayListItem oldItem, @NonNull WordSearchResultDayListItem newItem) {
            if (oldItem.getDayOfMonth() != newItem.getDayOfMonth()) {
                return false;
            }
            if (oldItem.getDayOfWeek() != null
                    && newItem.getDayOfWeek() != null
                    && !oldItem.getDayOfWeek().equals(newItem.getDayOfWeek())) {
                return false;
            }
            if (oldItem.getTitle() != null
                    && newItem.getTitle() != null
                    && !oldItem.getTitle().equals(newItem.getTitle())) {
                return false;
            }
            if (oldItem.getItemNumber() != newItem.getItemNumber()) {
                return false;
            }
            if (oldItem.getItemTitle() != null
                    && newItem.getItemTitle() != null
                    && !oldItem.getItemTitle().equals(newItem.getItemTitle())) {
                return false;
            }
            if (oldItem.getItemComment() != null
                    && newItem.getItemComment() != null
                    && !oldItem.getItemComment().equals(newItem.getItemComment())) {
                return false;
            }
            return true;
        }
    }
}

File: app/src/main/java/com/websarva/wings/android/zuboradiary/ui/adapter/diary/wordsearchresult/WordSearchResultDayListAdapter.kt
package com.websarva.wings.android.zuboradiary.ui.adapter.diary.wordsearchresult

import android.util.Log
import android.view.LayoutInflater
import android.view.ViewGroup
import androidx.recyclerview.widget.RecyclerView
import com.websarva.wings.android.zuboradiary.R
import com.websarva.wings.android.zuboradiary.domain.model.ThemeColor
import com.websarva.wings.android.zuboradiary.databinding.RowWordSearchResultListBinding
import com.websarva.wings.android.zuboradiary.ui.adapter.ListBaseAdapter
import com.websarva.wings.android.zuboradiary.utils.createLogTag
import com.websarva.wings.android.zuboradiary.ui.adapter.diary.DiaryDayListBaseDiffUtilItemCallback
import com.websarva.wings.android.zuboradiary.ui.adapter.diary.wordsearchresult.WordSearchResultDayListAdapter.WordSearchResultDayViewHolder
import com.websarva.wings.android.zuboradiary.ui.model.list.diary.wordsearchresult.WordSearchResultDayListItem
import com.websarva.wings.android.zuboradiary.ui.utils.toDiaryListDayOfWeekString
import java.text.NumberFormat

internal class WordSearchResultDayListAdapter(
    recyclerView: RecyclerView,
    themeColor: ThemeColor
) : ListBaseAdapter<WordSearchResultDayListItem, WordSearchResultDayViewHolder>(
    recyclerView,
    themeColor,
    DiffUtilItemCallback()
) {

    override fun build() {
        super.build()

        // MEMO:DiaryYearMonthListBaseAdapter#build()内にて理由記載)
        recyclerView.itemAnimator = null
    }

    override fun createViewHolder(
        parent: ViewGroup,
        themeColorInflater: LayoutInflater,
        viewType: Int
    ): WordSearchResultDayViewHolder {
        val binding =
            RowWordSearchResultListBinding.inflate(themeColorInflater, parent, false)
        return WordSearchResultDayViewHolder(binding)
    }

    override fun bindViewHolder(
        holder: WordSearchResultDayViewHolder,
        item: WordSearchResultDayListItem
    ) {
        holder.bind(
            item,
            themeColor
        ) { onClickItemListener?.onClick(it) }
    }

    class WordSearchResultDayViewHolder(
        val binding: RowWordSearchResultListBinding
    ) : RecyclerView.ViewHolder(binding.root) {

        fun bind(
            item: WordSearchResultDayListItem,
            themeColor: ThemeColor,
            onItemClick: (WordSearchResultDayListItem) -> Unit,
        ) {
            val context = binding.root.context
            binding.apply {
                val date = item.date
                val dayOfWeekString = date.dayOfWeek.toDiaryListDayOfWeekString(context)
                includeDay.textDayOfWeek.text = dayOfWeekString
                includeDay.textDayOfMonth.text = NumberFormat.getInstance().format(date.dayOfMonth)

                val title = item.createTitleSpannableString(context, themeColor)
                textTitle.text = title

                val strItemNumber =
                    context.getString(R.string.fragment_word_search_result_item) + item.itemNumber
                textItemNumber.text = strItemNumber
                textItemTitle.text = item.createItemTitleSpannableString(context, themeColor)
                textItemComment.text = item.createItemCommentSpannableString(context, themeColor)
            }
            itemView.setOnClickListener {
                onItemClick(item)
            }
        }
    }

    private class DiffUtilItemCallback :
        DiaryDayListBaseDiffUtilItemCallback<WordSearchResultDayListItem>() {

        private val logTag = createLogTag()

        override fun areContentsTheSame(
            oldItem: WordSearchResultDayListItem,
            newItem: WordSearchResultDayListItem
        ): Boolean {
            if (!oldItem.areContentsTheSame(newItem)) {
                Log.d(logTag, "areContentsTheSame()_不一致")
                return false
            }

            Log.d(logTag, "areContentsTheSame()_全項目一致")
            return true
        }
    }
}
